relay_address_generator_range.go

```diff
@@ -0,0 +1,105 @@
+package turn
+
+import (
+	"errors"
+	"fmt"
+	"github.com/pion/transport/vnet"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"time"
+)
+
+// RelayAddressGeneratorPortRange can be used to only allocate connections inside a defined port range.
+// Similar to the RelayAddressGeneratorStatic a static ip address can be set.
+type RelayAddressGeneratorPortRange struct {
+	// RelayAddress is the IP returned to the user when the relay is created
+	RelayAddress net.IP
+
+	// MinPort the minimum port to allocate
+	MinPort uint16
+	// MaxPort the maximum (inclusive) port to allocate
+	MaxPort uint16
+
+	// MaxRetries the amount of tries to allocate a random port in the defined range
+	MaxRetries int
+
+	// Rand the random source of numbers
+	Rand      *rand.Rand
+	randMutex sync.Mutex
+
+	// Address is passed to Listen/ListenPacket when creating the Relay
+	Address string
+
+	Net *vnet.Net
+}
+
+// Validate is called on server startup and confirms the RelayAddressGenerator is properly configured
+func (r *RelayAddressGeneratorPortRange) Validate() error {
+	if r.Net == nil {
+		r.Net = vnet.NewNet(nil)
+	}
+
+	if r.Rand == nil {
+		r.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	if r.MaxRetries == 0 {
+		r.MaxRetries = 10
+	}
+
+	switch {
+	case r.MinPort == 0:
+		return errors.New("MinPort must not be 0")
+	case r.MaxPort == 0:
+		return errors.New("MaxPort must not be 0")
+	case r.RelayAddress == nil:
+		return errRelayAddressInvalid
+	case r.Address == "":
+		return errListeningAddressInvalid
+	default:
+		return nil
+	}
+}
+
+func (r *RelayAddressGeneratorPortRange) randPort() uint16 {
+	r.randMutex.Lock()
+	defer r.randMutex.Unlock()
+	return r.MinPort + uint16(r.Rand.Intn(int((r.MaxPort+1)-r.MinPort)))
+}
+
+// AllocatePacketConn generates a new PacketConn to receive traffic on and the IP/Port to populate the allocation response with
+func (r *RelayAddressGeneratorPortRange) AllocatePacketConn(network string, requestedPort int) (net.PacketConn, net.Addr, error) {
+	if requestedPort != 0 {
```

Comment:
Is this needed? I've looked through the code and could only find requested port 0.

Reply:
This is used for `EVEN-PORT` and `RESERVATIONS`. This is never used in WebRTC TURN clients though.

I think this check makes sense!